bin/varnishd/cache/cache_deliver_proc.c

```diff
@@ -43,6 +43,13 @@
  * VDP_bytes will return that value directly without calling the next
  * processor.
  *
+ * VDP_END marks the end of successful processing, it is issued by
+ * VDP_DeliverObj() and may also be sent downstream by processors ending the
+ * stream (for return value != 0)
+ *
+ * VDP_END must at most be generated once per processor, so any VDP sending it
```

Comment:
This comment paragraph would be good in cache_filter.h -- make it part of the VDP API's public contract (and commit to staying with it, or else a major version change is necessary).

But does it mean that a VDP must not call VDP_bytes() with act=VDP_END, unless it is initiating the end of delivery? Or should (must?) a VDP push down VDP_END if it receives VDP_END? In which case, the VDP must be prepared to flush.

Unless I'm misunderstanding something, it has to be the latter, doesn't it? Otherwise it's not clear to me how lower-level VDPs know that we're at VDP_END.

Reply:
what I meant to say here is
* `VDP_END` must be the last thing pushed down
* every VDP must see at most one `VDP_END`. For error cases or premature ending further downstream, there can be no `VDP_END` before VDP close.

consequently,
* A VDP must only call `VDP_bytes()` with `act=VDP_END` with or after the last bytes
* after a VDP has pushed down `act=VDP_END`, it must ignore all other `.bytes` calls to itself

Thank you for pointing out that my documentation was lacking clarity. Does the above work for you?

And yes to moving it to cache_filter.h

Reply:
I would add, if this is the correct understanding: a VDP that receives VDP_bytes(act=VDP_END) with  MUST push down VDP_bytes(act=VDP_END) calls within that invocation.

Otherwise, it's mysterious to me how the lower-level VDPs know that we're at VDP_END.

Reply:
@slimhazard yes that or fail.